### PR TITLE
Add support for user stories

### PR DESCRIPTION
**Pull request by lampholder**
...et voilà

### CODE REVIEW

**Comment by turt2live** [comment]
image

File: src/App.js [hunk 1]
```diff
@@ -38,20 +40,35 @@ function getToken() {
     }
 }
 
+class FeatureDashboard {
+    static async getIssues(token, query) {
```
Comment (jryans) [inline src/App.js]: Why is this a separate class...? I am assuming the reason may reveal itself...

File: src/App.js [hunk 2]
```diff
@@ -38,20 +40,35 @@ function getToken() {
...
+                labels: []
+            },
+            issues: []
         }
     }
 
     async componentDidMount() {
+        let query = this.parseQueryFromHash(window.location.hash);
         this.setState({
             query: this.parseQueryFromHash(window.location.hash),
```
Comment (jryans) [inline src/App.js]: Could change to just `query,` since you've already calculated it above.

File: src/App.js [hunk 3]
```diff
@@ -42,11 +41,20 @@ function getToken() {
 
 class FeatureDashboard {
     static async getIssues(token, query) {
-        return await Github.getIssues(
-            token,
-            query.labels,
-            query.repos
-        );
+        if(query.epics) {
```
Comment (jryans) [inline src/App.js]: Nit: `if (`

Comment (jryans) [reply src/App.js]: (Maybe a good option to enable for ESLint...)

File: src/components/IssueTree.js [hunk 4]
```diff
@@ -31,45 +31,23 @@ class IssueTree extends Component {
...
-            let unbucketedItems = items.filter(item =>
-                !Object.values(buckets).reduce((array, value) => {
-                    return array.concat(value);
-                }, []).includes(item));
-
-            if (unbucketedItems.length > 0) {
-                buckets[unbucketed] = unbucketedItems;
-            }
-        }
+        let categorize = categories[0];
```
Comment (jryans) [inline src/components/IssueTree.js]: Whoa, US spelling! 😜

File: src/Github.js [hunk 5]
```diff
@@ -29,25 +29,35 @@ class Github {
...
             let label = `story:${userStory.number}`;
-            let storyIssues = await this.getFullIssues(token, [label], repos);
-            for (var storyIssue of storyIssues) {
-                storyIssue.story = userStory;
+            epic.push({
+                story: userStory,
+                relatedIssues: this.getFullIssues(token, [label], repos)
+            });
+        }
+        Promise.all(epic.map(userStory => userStory.issues));
```
Comment (jryans) [inline src/Github.js]: Hmm, I don't think `Promise.all` is actually doing anything here... Should it be removed?

File: src/Github.js [hunk 6]
```diff
@@ -267,16 +277,20 @@ class Github {
             }`
         let issues = [];
         // await Promise.all([]);
```
Comment (jryans) [inline src/Github.js]: Remove?

File: src/Github.js [hunk 7]
```diff
@@ -267,16 +277,20 @@ class Github {
...
+            return graphql(query, {
                 headers: {
                     authorization: `token ${token}`
                 },
                 owner: owner,
                 project: project,
                 labels: labels
             });
+        });
+        Promise.all(resultSets);
```
Comment (jryans) [inline src/Github.js]: You only need `Promise.all` if you want to `await` the entire array of promises. The actual work has already started (back when `graphql` was called). It should be safe to remove this, as other bits are ensuring all requests are issued in parallel.

File: src/Github.js [hunk 8]
```diff
@@ -267,16 +277,20 @@ class Github {
...
+        for (const resultSet of resultSets) {
+            let results = await resultSet;
```
Comment (jryans) [inline src/Github.js]: As discussed IRL, future work could make this trigger updates as each query returns, but that's for another day.

File: src/components/Plan.js [hunk 9]
```diff
@@ -22,7 +22,9 @@ import IssueTree from './IssueTree'
 const title = query => {
 
     if (query.epics) {
-        return query.epics.join(' ');
+        // FIXME: Inconsistent - can we speicify one epic or many?
+        console.log('titling', query.epics[0]);
```
Comment (jryans) [inline src/components/Plan.js]: Remove?

**Comment by jryans** [comment]
I have applied the review feedback and merged all of this **except** for the redesign bits, which we should consider separately. We can use https://github.com/vector-im/feature-dashboard/pull/56 for that.

**Pull request closed without merging**